core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java

```diff
@@ -461,16 +465,34 @@ private void decoratePillarWithClouderaManagerDatabase(Cluster cluster, Map<Stri
                 new SaltPillarProperties("/cloudera-manager/database.sls", singletonMap("cloudera-manager", singletonMap("database", rdsView))));
     }
 
-    private void decoratePillarWithClouderaManagerCommunicationSettings(Cluster cluster, Map<String, SaltPillarProperties> servicePillar) {
+    private void decoratePillarWithClouderaManagerCommunicationSettings(Stack stack, Cluster cluster,
+        Map<String, SaltPillarProperties> servicePillar,
+        GatewayConfig primaryGatewayConfig) {
         Boolean autoTls = cluster.getAutoTlsEnabled();
         Map<String, Object> communication = new HashMap<>();
+        Optional<LoadBalancer> loadBalancer = loadBalancerConfigService.getLoadBalancer(stack.getId(), LoadBalancerType.PRIVATE);
+        Optional<String> san = loadBalancer.flatMap(this::getBestSANForLB);
+        if (san.isPresent()) {
+            communication.put("internal_loadbalancer_san", san.get());
+        }
         communication.put("port", autoTls ? CM_HTTPS_PORT : CM_HTTP_PORT);
         communication.put("protocol", autoTls ? "https" : "http");
         communication.put("autotls_enabled", autoTls);
-        servicePillar.put("cloudera-manager-communication", new SaltPillarProperties("/cloudera-manager/communication.sls",
+        servicePillar.put("cloudera-manager-communication",
+            new SaltPillarProperties("/cloudera-manager/communication.sls",
                 singletonMap("cloudera-manager", singletonMap("communication", communication))));
     }
 
+    private Optional<String> getBestSANForLB(LoadBalancer lb) {
+        if (isNotBlank(lb.getFqdn())) {
+            return Optional.of("DNS:" + lb.getFqdn());
+        }
+        if (isNotBlank(lb.getDns())) {
+            return Optional.of("DNS:" + lb.getDns());
+        }
+        return Optional.of("IP:" + lb.getIp());
```

Comment:
Should we have a validation in place that the IP is indeed defined?

Reply:
No need, getLoadBalancer already filters out the load-balancers that does not have any of these.

```
    public Optional<LoadBalancer> getLoadBalancer(Long stackId, LoadBalancerType loadBalancerType) {
        return loadBalancerPersistenceService.findByStackId(stackId).stream()
            .filter(lb -> StringUtils.isNotBlank(lb.getDns()) ||
                StringUtils.isNotBlank(lb.getIp()) || StringUtils.isNotBlank(lb.getFqdn()))
            .filter(lb -> loadBalancerType.equals(lb.getType()))
            .findAny();
    }
```